treeherder/auth/backends.py

```diff
@@ -99,32 +95,40 @@ def authenticate(self, auth_header=None, host=None, port=None):
 
         if result["status"] != "auth-success":
             logger.warning("Error logging in: {}".format(result["message"]))
-            raise TaskclusterAuthenticationFailed(result["message"])
+            raise TaskclusterAuthenticationFailedException(result["message"])
 
+        client_id = result["clientId"]
+        email = self._get_email(client_id)
         # TODO: remove this size limit when we upgrade to django 1.10
         # in Bug 1311967
-        email = self._get_email(result)
-        if len(email) <= 30:
-            username = email
+        if len(client_id) <= 30:
+            username = client_id
         else:
             username = base64.urlsafe_b64encode(
-                hashlib.sha1(smart_bytes(email)).digest()
+                hashlib.sha1(smart_bytes(client_id)).digest()
                 ).rstrip(b'=')[-30:]
 
         try:
-            # Find the user by their email.
-            user = self._get_user(email)
-            user.username = username
+            if scope_match(result["scopes"],
+                           [["assume:mozilla-user:{}".format(email)]]):
+                # Find the user by their email.
+                user = self._get_user(email)
+                # update the username
+                user.username = username
+                user.save()
+                return user
 
-        except ObjectDoesNotExist:
-            # the user doesn't already exist, create it.
-            logger.warning("Creating new user: {}".format(username))
-            user = User(email=email,
-                        username=username,
-                        )
+            return User.objects.get(username=username)
 
-        user.save()
-        return user
+        except ObjectDoesNotExist:
+            if email:
+                # the user doesn't already exist, create it.
+                logger.warning("Creating new user: {}".format(username))
+                user = User.objects.create_user(email=email,
```

Comment:
Perhaps if we should use `.create()` rather than `create_user()` to avoid the need for the separate `.save()`?